lib/ddtrace/contrib/ethon/easy_patch.rb

```diff
@@ -0,0 +1,146 @@
+              remove_method :set_attributes
+
+              include InstanceMethods
+            end
+          else
+            base.send(:prepend, InstanceMethods)
+          end
+        end
+
+        # Compatibility shim for Rubies not supporting `.prepend`
+        module InstanceMethodsCompatibility
+          def set_attributes(options)
+            set_attributes_without_datadog(options)
+          end
+        end
+
+        # InstanceMethods - implementing instrumentation
+        module InstanceMethods
+          include InstanceMethodsCompatibility unless Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.0.0')
+
+          def http_request(url, action_name, options = {})
+            return super unless tracer_enabled?
+
+            # It's tricky to get HTTP method from libcurl
+            @datadog_method = action_name.to_s.upcase
+            super
+          end
+
+          def set_attributes(options)
+            return super unless tracer_enabled?
+
+            # Make sure headers= will get called
+            options[:headers] ||= {}
+            super options
+          end
+
+          def headers=(headers)
+            return super unless tracer_enabled?
+
+            # Store headers to call this method again when span is ready
+            headers ||= {}
+            @datadog_original_headers = headers
+            super headers
+          end
+
+          def perform
+            return super unless tracer_enabled?
+            datadog_before_request
+            super
+          end
+
+          def complete
+            return super unless tracer_enabled?
+            begin
+              response_options = mirror.options
+              response_code = (response_options[:response_code] || response_options[:code]).to_i
+              return_code = response_options[:return_code]
+              if response_code.zero?
+                message = return_code ? ::Ethon::Curl.easy_strerror(return_code) : 'unknown reason'
+                set_span_error_message("Request has failed: #{message}")
+              else
+                @datadog_span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response_code)
+                if Datadog::Ext::HTTP::ERROR_RANGE.cover?(response_code)
+                  set_span_error_message("Request has failed with HTTP error: #{response_code}")
+                end
+              end
+            ensure
+              @datadog_span.finish
+              @datadog_span = nil
+            end
+            super
+          end
+
+          def datadog_before_request
+            @datadog_span = datadog_configuration[:tracer].trace(
+              Ext::SPAN_REQUEST,
+              service: datadog_configuration[:service_name],
+              span_type: Datadog::Ext::HTTP::TYPE_OUTBOUND
+            )
+
+            datadog_tag_request
+
+            if datadog_configuration[:distributed_tracing]
+              Datadog::HTTPPropagator.inject!(@datadog_span.context, @datadog_original_headers)
```

Comment:
Why store the original headers then inject distributed tracing here? Could you `inject!` into `self.headers` without using an instance variable?

Reply:
I see; the basic thrust of my comment was to pursue simplicity, so if that suggestion wouldn't work or introduce more complexity, then I think this is fine.